requests-for-comments/2023-11-22-scheduled-flag-changes.md

```diff
@@ -0,0 +1,97 @@
+
+## Problem
+
+Currently, users can make updates to feature flags only manually and in real-time. This is not ideal, as it potentially leads to overlooked or forgotten changes, particularly when they fall outside normal working hours. This can lead to poorly timed releases, negatively impacting end users' engagement and feedback.
+
+## Solution
+
+Scheduling feature flag changes allows users to automatically change flag properties at a future point in time. For example, a user can schedule an initial release of the flag for internal testing, then enable it to the beta users group 2 days later, and finally roll it out to 100% of the users in 2 weeks.
+
+## Scope
+
+This feature should allow the following scheduled modification to a feature flag:
+
+-   **Enable/disable the flag** (field: `is_active`)
+-   **Add rollout conditions to the flag** (field: `filters`)
+
+TBD:
+
+-   **Modify existing rollout condition:** I would advise against adding this. When rolling out to more users, it is considered good practice to add to the existing conditions rather than modifying them. This provides an audit trail of changes and allows one to easily roll back the release to the previous state by removing the most recent condition(s).
+-   **Delete existing conditions:** This would allow temporary access to a feature. It might be helpful for some users, but I would advise to wait for more feedback before adding this.
+
+## Implementation
+
+We use Celery to execute tasks on schedule. At the predetermined time, Celery will pick the flag and evaluate scheduled changes. I considered two possible approaches to scheduling:
+
+-   Periodic scheduling - schedule every flag in regular intervals. This simplifies the implementation but creates a lot of unnecessary overhead.
+-   Dynamic scheduling - pick the earliest time from the scheduled changes and schedule the flag to be evaluated at this time. After performing the change, reschedule the flag to the new earliest scheduled time. For this RFC, I am assuming the use of dynamic scheduling.
+
+The scheduled changes can be stored on the flag record in the `scheduled_changes` field with the following structure:
+
+```
+[
+    {
+        "scheduled_at": "2023-11-22T12:30:00Z",
+        "type": "is_active",
+        "value": "false"
+    },
+    {
+        "scheduled_at": "2023-12-01T15:00:00Z",
+        "type": “filter”,
+        "value": {
+            "variant": null,
+            "properties": [{
+                "key": "$geoip_country_name",
+                "type": "person",
+                "value": ["Australia"],
+                "operator": "exact"
+            }],
+            "rollout_percentage": 20
+        }
+    }
+]
+```
+
+At the time specified, the Celery task fetches the flag, and iterates over the `scheduled_changes` items. For any item in this array, the `scheduled_at` field is evaluated, and if it is in the past, the task populates the item on the flag in the following way:
```

Comment:
This hides a lot of complexity, and I'm not sure it's possible the way we intend.

1. Our current celery backend is redis, which doesn't guarantee tasks will stick around
2. Celery workers will restart on every deploy
3. Celery recommends not scheduling tasks a long time in the future (like more than a few minutes): https://docs.celeryq.dev/en/stable/userguide/calling.html#eta-and-countdown
  and I'm not sure the alternative provided here will work in our case.
  
So, I'm leaning towards periodic checks to see if there's something to be scheduled every minute or so, this will have minimal overhead if we design things right (like a table with index on scheduled_time)

Alternatively, I think it's worth checking with @PostHog/team-pipeline what's the state of temporal, if it makes sense for this use case, & if we're planning to move everything to temporal at all or not. Not 100% clear of the capabilities there but if it's good enough for batch exports running at specific times, might be just what we need? or if it's way too heavy & overkill for what we need

Reply:
Agreed, Celery is designed for repeated tasks than can be harmlessly skipped (as they'll be re-scheduled again). Temporal could indeed be a good scheduler for it, but might be a heavier lift.

Could you structure the database schema so that an idempotent "get transitions to execute and update the relevant stuff" transaction could be run? That transaction could then run every N minutes on Celery without reliability concerns. Happy to huddle on that.

Reply:
> Could you structure the database schema so that an idempotent "get transitions to execute and update the relevant stuff" transaction could be run? 

Yes, 100%! - I do currently like this the most as well.

Reply:
FYI, pipeline is currently designing webhooks v2 that will be based on a custom Postgres task queue, so we'll be able to assist with concurrency when that becomes a scalability concern.